### ISSUE TITLE
Route through squares (area=yes)

### ISSUE THREAD

**Issue by emiltin** [Project-OSRM/osrm-backend#64]
handling square might not be trivial, and are probably mostly important for pedestrian routing. so this might be for a future version.

the thing with a square is that you can cross it in any way. unfortunately squares can have odd shapes, so it's not always possible go in a straight line from where you enter to where you leave. for example a building might be in the way if the square has a hour-glass shape.

a simple example is "Prins Jørgens Gård" is a square, tagged with highway=pedestrian, area=yes:
http://www.openstreetmap.org/?lat=55.676763&lon=12.579427&zoom=18&layers=M

if i set up a pedestrain speedprofile, which includes 'pedestrian=5', osrm will route at the edge of the square. this is will get me across the square, but in a somewhat odd fashion.

one option would be to add routes connecting all entries to the square to all others. another would be to connect all entries to a central point. perhaps these to options could be improved by checking that routes are only added if they don't cross the square boundary.

a third option would be be still route along the edge of the square, but insert the route some reasonably distance from the edge, since in reality you cant walk right at the edge of the building.

**Comment by TheMarex** [comment]
Brain dump on how to do this:
- This requires that we have all node locations available when processing a way
- Add area flag to ExtractionWay
- In ExtractorCallbacks::ProcessWay compute the visibility graph of the polygon:
  - For each pair of boundary vertices check if they can be connected without intersecting the boundary
  - Add edge for pair if intersection free

**Comment by lbud** [comment]
this would also be awesome for, say, indoor routing :)

**Comment by danpat** [comment]
Example indoor routing use-case:  http://en.wikipedia.org/wiki/Edmonton_Pedway

Some of the walkways are in OSM, but it doesn't look complete:  https://www.openstreetmap.org/#map=17/53.54259/-113.49257

**Comment by homersimpsons** [comment]
Is it possible, in preprocessing operation, to create ways in those area?

**Comment by daniel-j-h** [comment]
Yes, you could do the pre-processing already on the OSM extract adding artificial ways and saving out the extract again. OSRM will then happily route you over those ways when appropriate.

Have a look at https://github.com/osmcode/libosmium if you want to give it a try.

**Comment by daniel-j-h** [comment]
Let's be more concrete here. What needs to be done is the following.

- Filter [`area=yes`](http://wiki.openstreetmap.org/wiki/Key:area) + [`highway=pedestrian`](http://wiki.openstreetmap.org/wiki/Key:highway) ways. In addition [access](http://wiki.openstreetmap.org/wiki/Key:access) tags need to be respected as well. Open question: are there other tags / combinations we have to take care of? Sample OSM!
- Handle simple areas (no [multipolygones](http://wiki.openstreetmap.org/wiki/Relation:multipolygon)) at first / at all. Osmium's `Area` type - the [`area()` handler callback](http://docs.osmcode.org/libosmium/latest/classosmium_1_1handler_1_1Handler.html) is getting called with - has a `is_multipolygon()` check.
- Compute visibility graph. Simple solution as outlined above is good enough for a first pass. We can make it fancy afterwards. Care needs to be taken not to blow up the already huge walk graph. Gather some statistics for the planet how many areas there are and how many support nodes they have min. / max. / avg. to get a feeling for the data out there.


We could do it either as a pre-processing step as noted above. Implementing it internally has the benefit of us being able to tell artificial ways we introduce apart from original OSM data.

Additional work could then propagate a flag through the pipeline allowing us to "fix" guidance by introducing "cross the area" instructions.

Internally we would need to classify areas based on the raw OSM objects [here](https://github.com/Project-OSRM/osrm-backend/blob/bf6698f4cc7608a75d9ac72f18248356b6ed162c/src/extractor/extractor.cpp#L205-L222) and create a new `ExtractorCallbacks::ProcessArea` callback taking a area and inserting artificial ways (marked as such).

Ref. refactoring the extraction callbacks for proper libosmium usage: https://github.com/Project-OSRM/osrm-backend/issues/3447


cc @PedaB @joto

**Comment by joto** [comment]
@daniel-j-h Why no multipolygons? Are they more difficult to handle?

**Comment by daniel-j-h** [comment]
Yes, at least the simple approach won't do for multipolygons. Have a look at [these examples](http://wiki.openstreetmap.org/wiki/Multipolygon_Examples#Forest_with_lake_.28One_outer_and_one_inner_ring.29) in the context of area routing via creating artificial ways.

**Comment by joto** [comment]
@daniel-j-h Note that the `osmium::Area::is_multipolygon()` call will only check whether the area has more than one outer ring. What you seem to be interested in is that it doesn't have any inner rings. You'd have to call `num_rings()` for that.

**Comment by emiltin** [comment]
regarding tag combinations, we should reuse as much of the usual way processing as possible, to handle tags like access, foot, bicycle, car, surface, smoothness, construction, impassable, status, amenity, platform, railway, service, steps, etc.

tags related to direction will not apply to squares, like oneway, *:forward and *:backward and should be ignored (and would be invalid tagging anywys). same with tags related to lanes and width.

turn restrictions to/from a square probably doesn't apply either.

**Comment by sfkeller** [comment]
What is the reason that this issue about pedestrian routing through areas has such low priority?  

I'm asking because mappers obviously continue to add "virtual" ways over areas in OSM. This is unncessary and like "tagging for the router". In addition it looks weird in renderers, because such pedestrian ways can't be distinguished from "real" ones without guessing...

**Comment by daniel-j-h** [comment]
Mainly two reasons
1. in the last two years the core dev team heavily focused on car navigation;
    and @emiltin is working on all things bike routing for his own use-case
2. no one stepped up and wanted to implement it;
    if you have a use-case for this and see a need here we're more than happy to guide you along.

**Comment by emiltin** [comment]
This is highly relevant for bicycle routing as well (which is why I opened the issue originally), since:

1) you're allowed to bicycle on some areas
2) you can push your bike on most areas where you're allowed to walk

But unfortunately I don't have time to implement it.

**Comment by systemed** [comment]
FWIW, it's not "tagging for the renderer/router". That means putting _incorrect_ information in to achieve the desired effect: for example, tagging a golf course bunker as a beach so it shows up as sand, even though it's not a beach. Adding centrelines or virtual ways is perhaps duplicate information, but it's been done since the very earliest days of OSM (on rivers).

**Comment by andrewharvey** [comment]
> Open question: are there other tags / combinations we have to take care of? Sample OSM!

[`natural=beach`](https://wiki.openstreetmap.org/wiki/Tag:natural=beach) for the walking profile. Beaches don't normally have a path going through them but you can walk along them.

**Comment by sfkeller** [comment]
@systemed: I'm talking about not-existing and misleading (sic!) footways which have been added only to enhance a router, like on this place "Helvetiaplatz, Zurich" (http://www.openstreetmap.org/way/4533221 ). There you also see that these unnecessary footways are only partially working since the route from east to west side is missing... 
![capture](https://user-images.githubusercontent.com/43340/31353574-61daecee-ad33-11e7-897d-af51941d16f1.PNG)

**Comment by saerdnaer** [comment]
> if you have a use-case for this and see a need here we're more than happy to guide you along.
@daniel-j-h  Can you give some pointers how to start implementing this issue? Which classes have to be modified? (besides https://github.com/Project-OSRM/osrm-backend/issues/64#issuecomment-289463048 )

**Comment by daniel-j-h** [comment]
If you give the `ExtractionWay` [here](https://github.com/Project-OSRM/osrm-backend/blob/3b4e2e83ef85983df1381dbeacd0ea5d4b9bbbcb/include/extractor/extraction_way.hpp#L37) a new property maybe `is_area` and expose it to the profiles through the scripting environment [here](https://github.com/Project-OSRM/osrm-backend/blob/3b4e2e83ef85983df1381dbeacd0ea5d4b9bbbcb/src/extractor/scripting_environment_lua.cpp#L309) you can parse a way's area tag in the Lua profiles and set the attribute accordingly.

You probably also want to expose `is_closed` on the [osmium::Way](http://docs.osmcode.org/libosmium/latest/classosmium_1_1Way.html#af492debe7ea1e776fe32e42631d603c2) in the scripting environment [here](https://github.com/Project-OSRM/osrm-backend/blob/3b4e2e83ef85983df1381dbeacd0ea5d4b9bbbcb/src/extractor/scripting_environment_lua.cpp#L253-L271) because you only care for closed ways (=> areas).

Once you have the attribute on the ExtractionWay and inside C++-land you can modify the ExtractorCallbacks [here](https://github.com/Project-OSRM/osrm-backend/blob/3b4e2e83ef85983df1381dbeacd0ea5d4b9bbbcb/src/extractor/extractor_callbacks.cpp#L92) to compute e.g. the visibility graph and add edges.

Hope that helps

**Comment by dreirund** [comment]
*See also #4176 and #7146.*

**Comment by emiltin** [comment]
nice to see this ancient issue completed 👍
